subsys/mgmt/hawkbit/hawkbit.c

```diff
@@ -1197,240 +1047,458 @@ static bool check_hawkbit_server(void)
+		LOG_DBG("config.polling.sleep=%s", s->hb_context.results.base.config.polling.sleep);
+		hawkbit_update_sleep(&s->hb_context.results.base);
 	}
 
+	if (s->hb_context.results.base._links.cancelAction.href) {
+		LOG_DBG("_links.%s.href=%s", "cancelAction",
+			s->hb_context.results.base._links.cancelAction.href);
+		smf_set_state(SMF_CTX(s), &hawkbit_states[S_HAWKBIT_CANCEL]);
+		return;
+	}
 
-	if (hawkbit_results.base._links.cancelAction.href) {
-		ret = hawkbit_find_cancelAction_base(&hawkbit_results.base, cancel_base);
-		memset(hb_context.url_buffer, 0, sizeof(hb_context.url_buffer));
-		hb_context.dl.http_content_size = 0;
-		snprintk(hb_context.url_buffer, sizeof(hb_context.url_buffer),
-			 "%s/%s-%s/%s/feedback", HAWKBIT_JSON_URL, CONFIG_BOARD, device_id,
-			 cancel_base);
-		memset(&hawkbit_results.cancel, 0, sizeof(hawkbit_results.cancel));
-
-		if (!send_request(HTTP_POST, HAWKBIT_CLOSE, HAWKBIT_STATUS_FINISHED_SUCCESS,
-				  HAWKBIT_STATUS_EXEC_CLOSED)) {
-			LOG_ERR("Send request failed (%s)", "HAWKBIT_CLOSE");
-			hb_context.code_status = HAWKBIT_NETWORKING_ERROR;
-			goto cleanup;
-		}
+	if (s->hb_context.results.base._links.configData.href) {
+		LOG_DBG("_links.%s.href=%s", "configData",
+			s->hb_context.results.base._links.configData.href);
+		smf_set_state(SMF_CTX(s), &hawkbit_states[S_HAWKBIT_CONFIG_DEVICE]);
+		return;
+	}
 
-		hb_context.code_status = HAWKBIT_CANCEL_UPDATE;
-		goto cleanup;
+	if (s->hb_context.results.base._links.deploymentBase.href) {
+		LOG_DBG("_links.%s.href=%s", "deploymentBase",
+			s->hb_context.results.base._links.deploymentBase.href);
+		smf_set_state(SMF_CTX(s), &hawkbit_states[S_HAWKBIT_PROBE_DEPLOYMENT_BASE]);
+		return;
 	}
 
-	if (hawkbit_results.base._links.configData.href) {
-		LOG_DBG("_links.%s.href=%s", "configData",
-			hawkbit_results.base._links.configData.href);
-		memset(hb_context.url_buffer, 0, sizeof(hb_context.url_buffer));
-		hb_context.dl.http_content_size = 0;
-		snprintk(hb_context.url_buffer, sizeof(hb_context.url_buffer), "%s/%s-%s/%s",
-			 HAWKBIT_JSON_URL, CONFIG_BOARD, device_id, "configData");
-
-		if (!send_request(HTTP_PUT, HAWKBIT_CONFIG_DEVICE, HAWKBIT_STATUS_FINISHED_SUCCESS,
-				  HAWKBIT_STATUS_EXEC_CLOSED)) {
-			LOG_ERR("Send request failed (%s)", "HAWKBIT_CONFIG_DEVICE");
-			hb_context.code_status = HAWKBIT_NETWORKING_ERROR;
-			goto cleanup;
-		}
+	s->hb_context.code_status = HAWKBIT_NO_UPDATE;
+	smf_set_state(SMF_CTX(s), &hawkbit_states[S_HAWKBIT_TERMINATE]);
+}
+
+/*
+ * Feedback channel for cancel actions
+ * POST: /{tenant}/controller/v1/{controllerId}/cancelAction/{actionId}/feedback
+ */
+static void s_cancel(void *o)
+{
+	int ret = 0;
+	int32_t cancel_action_id = 0;
+	struct s_object *s = (struct s_object *)o;
+	uint8_t status_buffer[CONFIG_HAWKBIT_STATUS_BUFFER_SIZE] = {0};
+	char *url_buffer;
+
+	url_buffer = hawkbit_get_url(s->hb_context.results.base._links.cancelAction.href);
```

Comment:
It seems like the resulting url is missing the "actionId" and the ending "feedback".

I believe that "actionId" is unintentionally removed by hawkbit_find_cancel_action_id which uses strtok, which in turn is modifying the str in place. A solution would be to copy href to a buffer before modifying it.